File: docs/DSA/src/py/time_based_key_value_store.py
from collections import defaultdict


class TimeMap:

    def __init__(self):
        self.data = defaultdict(list)


    def set(self, key: str, value: str, timestamp: int) -> None:
        self.data[key].append((value, timestamp))

    def get(self, key: str, timestamp: int) -> str:
        items = self.data[key]
        if not items: return ""
        left, right = 0, len(items)
        while left < right:
            mid = left + (right-left)//2
            if timestamp < items[mid][1]:
                right = mid
            else:
                left = mid + 1
        return items[left - 1][0] if left > 0 else ""

File: docs/DSA/src/go/time_based_key_value_store.go
package main

type Datapoint struct {
	value     string
	timestamp int
}

type TimeMap struct {
	data map[string][]Datapoint
}

func Constructor2() TimeMap {
	return TimeMap{
		data: make(map[string][]Datapoint),
	}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	this.data[key] = append(this.data[key], Datapoint{
		value:     value,
		timestamp: timestamp,
	})
}

func (this *TimeMap) Get(key string, timestamp int) string {
	items, ok := this.data[key]
	if !ok {
		return ""
	}
	left, right := 0, len(items)
	var mid int
	for left < right {
		mid = left + (right-left)/2
		if timestamp < items[mid].timestamp {
			right = mid
		} else {
			left = mid + 1
		}
	}
	if left > 0 {
		return items[left-1].value
	}
	return ""
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */
